ydb/core/tx/schemeshard/schemeshard_build_index__progress.cpp

```diff
@@ -261,6 +267,138 @@ THolder<TEvSchemeShard::TEvModifySchemeTransaction> InitiatePropose(
+    TSchemeShard* ss, const TIndexBuildInfo& buildInfo)
+{
+    Y_ASSERT(buildInfo.IsBuildVectorIndex());
+
+    auto propose = MakeHolder<TEvSchemeShard::TEvModifySchemeTransaction>(ui64(buildInfo.ApplyTxId), ss->TabletID());
+    propose->Record.SetFailOnExist(true);
+
+    auto path = TPath::Init(buildInfo.TablePathId, ss);
+    auto tableId = path->PathId;
+    path.Dive(buildInfo.IndexName);
+
+    NKikimrSchemeOp::TModifyScheme& modifyScheme = *propose->Record.AddTransaction();
+    modifyScheme.SetInternal(true);
+    modifyScheme.SetWorkingDir(path.PathString());
+
+    using namespace NTableIndex::NTableVectorKmeansTreeIndex;
+    TString name0 = PostingTable;
+    TString name1 = PostingTable;
+    const char* suffix0 = buildInfo.KMeans.Level % 2 == 0 ? TmpPostingTableSuffix0 : TmpPostingTableSuffix1;
+    const char* suffix1 = buildInfo.KMeans.Level % 2 == 0 ? TmpPostingTableSuffix1 : TmpPostingTableSuffix0;
+    name0.append(suffix0);
+    name1.append(suffix1);
+
+    // TODO(mbkkt) for levels greater than zero we need to disable split/merge completely
+    // For now it's not guranteed, but very likely
+    // But lock is really unconvinient approach (needs to store TxId/etc)
+    // So maybe best way to do this is specify something in defintion, that will prevent these operations like IsBackup
+
+    modifyScheme.SetOperationType(NKikimrSchemeOp::ESchemeOpInitiateBuildIndexImplTable);
+    auto& op = *modifyScheme.MutableCreateTable();
+
+    op = ss->Tables.at(path.Dive(name1)->PathId)->TableDescription;
+    op.ClearPath();
+    op.SetName(name0);
+
+    THashMap<ui32, TString> keyColumns;
+    for (auto& column : *op.MutableColumns()) {
+        column.ClearFamily();
+        column.ClearFamilyName();
+        auto [typeInfo, typeMod] = NScheme::TypeInfoModFromProtoColumnType(column.GetTypeId(), &column.GetTypeInfo());
+        column.SetType(NScheme::TypeName(typeInfo, typeMod));
+        keyColumns.emplace(column.GetId(), column.GetName());
+    }
+    op.ClearKeyColumnNames();
+    for (auto id : op.GetKeyColumnIds()) {
+        op.AddKeyColumnNames(keyColumns[id]);
+    }
+
+    op.SetSystemColumnNamesAllowed(true);
+
+    const auto& kmeans = buildInfo.KMeans;
+    Y_ASSERT(kmeans.K != 0);
+    Y_ASSERT((kmeans.K & (kmeans.K - 1)) == 0);
+    auto pow = [](ui32 k, ui32 l) {
+        ui32 r = 1;
+        while (l != 0) {
+            if (l % 2 != 0) {
+                r *= k;
+            }
+            k *= k;
+            l /= 2;
+        }
+        return r;
+    };
+    const auto count = pow(kmeans.K, kmeans.Level + 1);
+    auto step = 1;
+    auto parts = count;
+    auto shards = ss->Tables.at(tableId)->GetShard2PartitionIdx().size();
+    if (buildInfo.KMeans.Level + 1 == buildInfo.KMeans.Levels || shards <= 1) {
+        shards = 1;
+        parts = 1;
+    }
+    for (; shards < parts; parts /= 2) {
+        step *= 2;
+    }
+    for (; parts < shards / 2; parts *= 2) {
+        Y_ASSERT(step == 1);
+    }
```

Comment:
Would it be possible to separate protobuf message filling from the calculation of various values needed to fill the message with?

Reply:
Done